CurrencyConverterControl/Data/CurrencyDataProvider.cs

```diff
@@ -0,0 +1,164 @@
+
+                    if (response.IsSuccessStatusCode)
+                    {
+                        string content =  response.Content.ReadAsStringAsync().Result;
+                        JObject json = JObject.Parse(content);
+                        double value = (double)json["result"];                        
+                        return value;
+                    }
+                    else
+                    {
+                        Console.WriteLine("Request failed with status code: " + response.StatusCode);
+                    }
+                
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine($"Error: {ex.Message}");
+            }
+            return -1;
+        }
+        /// <summary>
+        /// This method is used to get List of currencies which holds Currency Code and Flag 
+        /// </summary>
+        /// <returns>Returns a List of Currency Objects </returns>
+        /// <exception cref="DirectoryNotFoundException">When Path of the file is not found</exception>
+        public List<Currency> GetCurrenciesData()
+        {
+            string filePath = @"Resources\CurrencyListXML.xml";
+
+            // Call the method to read data
+            return DeserializeFromXml(filePath);
+        }
+        private List<Currency> DeserializeFromXml(string inputPath)
+            {
+               CurrencyList currencies = new CurrencyList();
+
+                XmlSerializer serializer = new XmlSerializer(typeof(CurrencyList));
+
+                try
+                {
+                    using (TextReader reader = new StreamReader(inputPath))
+                    {
+                        currencies = (CurrencyList)serializer.Deserialize(reader);
+                    }
+
+                    Console.WriteLine($"XML file successfully deserialized from: {inputPath}");
+                }
+                catch (Exception ex)
+                {
+                    Console.WriteLine($"Error deserializing XML file: {ex.Message}");
+                }
+
+                return currencies.CurrenciesList;
+            }
+
+        /// <summary>
+        /// Gets the list of convertion Rates for two currencies
+        /// </summary>
+        /// <param name="currencyfrom"></param>
+        /// <param name="currencyto"></param>        
+        /// <returns>Returns the value of the convertion at that moment</returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        /// <exception cref="System.NullReferenceException">Happens when your monthly usage limit has been reached.  </exception>
+        public double GetConversionRate(string currencyfrom, string currencyto)
+        {
+            try
+            {
+                if (exchangeRates.Count > 0)
+                {
+                    string parameterValue = String.Concat(currencyfrom, currencyto);
+                    return exchangeRates.GetValueOrDefault(parameterValue);
```

Comment:
you store only the exchange rates with the source in USD in the exchangeRates.
If currencyfrom is another currency, nothing will be found. 
You need to get rates for USD=>currencyfrom and USD=>currencyto and calculate ratio currencyfrom=>currencyto using this data and return it.

Reply:
Yes, based on the requirements that for the DataGrid will perform only exchanges from USD to 'x' currency i have found that API which holds the exchange rates only for what l needed. That is why l chose that solution.

Reply:
Yes. But the DataGrid app is just an example. When you design a control, you don't want to use it in just one case. It should work in other cases as well. And now you have a working public API - ConvertValues.

What happens if user call ConvertValues with currencyFrom argument other then USD?